### ISSUE TITLE
 Cross references (target-counter) do not work for backward references

### ISSUE THREAD

**Issue by mattieusch** [Kozea/WeasyPrint#786]
Target-counter references work great for TOC generation. TOCs are usually in the beginning of the document, therefore target-counters usually reference elements 'later on' in the document (which works fine).

However, there seems to be a bug when target-counters reference elements that were already processed. See example below. The TOC on the first page renders correctly, the TOC on the last page has '0'-s for the page number (tested with WeasyPrint 44). This is especially problematic for cross references (which, depending on the position in the document of the target, may or may not work correctly).

```
<html>
<head>
    <style>
    .toc a::after {
      content: " .... " target-counter(attr(href), page);
    }    
    div.page {
        page-break-after:always;
    }
    @page {
        @top-center {
            content: "Page " counter(page) " of " counter(pages);
        }
    }
    </style>
</head>
<body>
    <div class="page toc">
      <a href="#h1">To header 1</a><br>
      <a href="#h2">To header 2</a><br>
      <a href="#h3">To header 3</a><br>
    </div>

    <div class="page">
      <h1 id="h1">Header 1</h1>
    </div>
    <div class="page">
      <h1 id="h2">Header 2</h1>
    </div>
    <div class="page">
      <h1 id="h3">Header 3</h1>        
    </div>

    <div class="page toc">
      <a href="#h1">To header 1</a><br>
      <a href="#h2">To header 2</a><br>
      <a href="#h3">To header 3</a><br>
    </div>
  </body>
</html>
```
Observation: in a rather large test document I did notice that in SOME cases the 'back referencing' did work. I strongly suspect that this was due to the fact that the target elements that had valid page counter numbers were being repositioned to a new page during layout. (Perhaps this triggers a different handling for cached_page_counter_values??). I would love to have solved this issue myself, but I am not very proficient with Python unfortunately.

**Comment by Tontyna** [comment]
The question is:  Why does  the *back reference* / the second TOC believe its page based target counters being up-to-date whilst they aren't? 
And: Why do some back references sometimes get their targets right?

I'll need a while to familiarize myself with the logic behind the TargetCollector again...

**Comment by mattieusch** [comment]
Hi Tontyna - thanks! I can confirm that the fix in #788 indeed solves the backref issues. Thanks!
